src/ucp/wireup/wireup.c

```diff
@@ -1355,25 +1355,184 @@ static void ucp_wireup_discard_uct_eps(ucp_ep_h ep, uct_ep_h *uct_eps,
+                           const ucp_tl_bitmap_t *tl_bitmap,
+                           const ucp_unpacked_address_t *remote_address)
+{
+    ucp_ep_config_key_t key;
+    unsigned addr_indices[UCP_MAX_LANES];
+
+    ucp_ep_config_key_reset(&key);
+
+    /* Perform selection in order to initialize configuration key */
+    if (ucp_wireup_select_lanes(ep, ep_init_flags, *tl_bitmap, remote_address,
+                                addr_indices, &key, 1) != UCS_OK) {
+        return 0;
+    }
+
+    return ucp_wireup_can_reconfigure_internal(ep, &key, remote_address,
+                                               addr_indices);
+}
+
+static ucp_lane_index_t
+ucp_wireup_find_non_reused_lane(ucp_ep_h ep, const ucp_ep_config_key_t *key)
+{
+    if (ucp_ep_has_cm_lane(ep)) {
+        return key->cm_lane;
+    }
+
+    /* Just use first lane, as only non-reused lanes are allowed at the
+     * moment. */
+    ucs_assert(key->num_lanes > 0);
+    return 0;
+}
+
+static ucs_status_t ucp_wireup_replace_wireup_msg_lane(ucp_ep_h ep,
+                                                       ucp_ep_config_key_t *key,
+                                                       uct_ep_h *new_uct_eps)
+{
+    uct_ep_h uct_ep = NULL;
+    ucp_lane_index_t old_lane, new_lane;
+    ucp_wireup_ep_t *old_ep, *new_ep;
+    uct_ep_h msg_ep;
+    ucp_rsc_index_t aux_rsc_index;
+    ucs_queue_head_t pending_queue;
+    int is_p2p;
+    ucs_status_t status;
+    ucp_request_t *req;
+
+    /* Get old wireup lane */
+    old_lane = ucp_wireup_get_msg_lane(ep, UCP_WIREUP_MSG_REQUEST);
+    old_ep   = ucp_wireup_ep(ucp_ep_get_lane(ep, old_lane));
+    ucs_assert(old_ep != NULL);
+
+    /* Set wireup EP for new configuration's wireup lane */
+    if (ucp_ep_has_cm_lane(ep)) {
+        /* Use existing EP from CM lane */
+        new_ep = ucp_ep_get_cm_wireup_ep(ep);
+        ucs_assert(new_ep != NULL);
+        aux_rsc_index = ucp_ep_get_rsc_index(ep, old_lane);
+        is_p2p        = ucp_ep_is_lane_p2p(ep, old_lane);
+    } else {
+        /* Create new EP for non-CM flow */
+        status = ucp_wireup_ep_create(ep, &uct_ep);
+        if (status != UCS_OK) {
+            return status;
+        }
+        new_ep        = ucp_wireup_ep(uct_ep);
+        aux_rsc_index = ucp_wireup_ep_get_aux_rsc_index(&old_ep->super.super);
+        is_p2p        = old_ep->flags & UCP_WIREUP_EP_FLAG_AUX_P2P;
+    }
+
+    ucs_assert(aux_rsc_index != UCP_NULL_RESOURCE);
+    ucs_queue_head_init(&pending_queue);
+
+    /* Move aux EP to new wireup lane */
+    msg_ep = ucp_wireup_ep_extract_msg_ep(old_ep, &pending_queue);
+    ucp_wireup_ep_set_aux(new_ep, msg_ep, aux_rsc_index, is_p2p);
+
+    /* Add wireup messages to pending queue */
+    ucs_queue_for_each_extract(req, &pending_queue, send.uct.priv, 1) {
+        status = uct_ep_pending_add(&new_ep->super.super, &req->send.uct, 0);
+        if (status != UCS_OK) {
+            goto destroy_ep;
```

Comment:
rest of pending queue wont be handled in case of error

Reply:
why is it relevant?

Reply:
in case of error it leads to memory leak and probably hang due to uncompleted lost requests